sematic/resolvers/local_resolver.py

```diff
@@ -446,9 +505,27 @@ def _populate_graph(
         Update the graph based on future.
         """
         if future.id not in self._runs:
+            # TODO: the run is being created as a side effect of updating the graph;
+            # the run and future should be atomically updated, if not the same concept;
+            # the run lifecycle should be the main flow the resolver deals with, and the
+            # graph should be a consequence of this, without being updated through lateral
+            # effect such as the run is now, and without the disjointed partial save from
+            # below
             run = self._make_run(future)
             self._add_run(run)
 
+            # attempt to directly resolve the run based on previous executions
+            artifact = self._get_artifact_from_cache(run=run)
+            if artifact is not None:
+                run.future_state = FutureState.RESOLVED
+                # in order to update the future's value below, its run must exist in the
+                # DB, so we save it here; subsequent refinements will add edges and such
+                self._save_graph()
+                # TODO: here we have to deserialize the type and value just to serialize
+                # it again; there should be a short-circuit
```

Comment:
Not sure such an optimization is worth it; it costs us maybe on the order of seconds max, whereas each step in the pipeline will generally take at least a few seconds to spin up/down the compute node for. Luckily we are not aiming for this DAG framework to be a high-performance in-memory system as well as an orchestration system 😆 .

Reply:
Removed.